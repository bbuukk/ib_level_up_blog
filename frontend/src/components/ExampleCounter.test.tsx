import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Counter from './ExampleCounter';
import { renderRoute } from 'tests/wrappers';

describe('Counter', () => {
  it('renders initial count and increments when button is clicked', () => {
    render(<Counter />);

    // Check initial render
    // getByText: RTL query
    // toBeInTheDocument: jest-dom matcher
    expect(screen.getByText('Count: 0')).toBeInTheDocument();

    // Click button and check if count incremented
    const button = screen.getByText('Increment');
    // fireEvent : RTL API
    fireEvent.click(button);
    expect(screen.getByText(/count: 1/i)).toBeInTheDocument();
    expect(screen.queryByText(/count: 0/i)).not.toBeInTheDocument();
  });
  it('updates the value of the input when typing on it', () => {
    render(<Counter />);

    const input = screen.getByLabelText('Name:');
    fireEvent.change(input, { target: { value: 'John Doe' } });
    expect(input).toHaveValue('John Doe');
  });
  it('typing in an input - with userEvent', async () => {
    const user = userEvent.setup();
    render(<Counter />);

    const textbox = screen.getByLabelText('Name:');
    const testDataTestId = screen.getByTestId('name-input');
    const testRole = screen.getByRole('textbox', { name: 'Name:' });
    await user.type(textbox, 'Hello');

    //screen.debug(undefined, Infinity);

    expect(textbox).toHaveValue('Hello');
  });

  it('test counter in app', async () => {
    const container = renderRoute({
      memoryRouterOptions: { initialEntries: ['/playground'] }
    });

    // Check initial render
    expect(await container.findByText('Count: 0')).toBeInTheDocument();

    // Click button and check if count incremented
    const button = container.getByText('Increment');
    fireEvent.click(button);
    expect(container.getByText('Count: 1')).toBeInTheDocument();
  });
});
