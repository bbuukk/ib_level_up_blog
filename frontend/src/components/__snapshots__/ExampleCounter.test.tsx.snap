// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Counter > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="p-4"
      >
        <p
          class="text-lg font-semibold"
        >
          Count: 
          0
        </p>
        <button
          class="mt-2 block rounded bg-blue-500 px-4 py-2 font-bold text-white hover:bg-blue-700"
        >
          Increment
        </button>
        <div
          class="mt-4"
        >
          <label
            class="block text-sm font-medium text-white"
            for="name"
          >
            Name:
          </label>
          <input
            class="mt-1 block rounded-md border-gray-700 bg-gray-800 text-white shadow-sm focus:border-indigo-500 focus:ring-indigo-500 sm:text-sm"
            data-testid="name-input"
            id="name"
            placeholder="Type your name"
            type="text"
            value=""
          />
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="p-4"
    >
      <p
        class="text-lg font-semibold"
      >
        Count: 
        0
      </p>
      <button
        class="mt-2 block rounded bg-blue-500 px-4 py-2 font-bold text-white hover:bg-blue-700"
      >
        Increment
      </button>
      <div
        class="mt-4"
      >
        <label
          class="block text-sm font-medium text-white"
          for="name"
        >
          Name:
        </label>
        <input
          class="mt-1 block rounded-md border-gray-700 bg-gray-800 text-white shadow-sm focus:border-indigo-500 focus:ring-indigo-500 sm:text-sm"
          data-testid="name-input"
          id="name"
          placeholder="Type your name"
          type="text"
          value=""
        />
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
